sopel/tools/jobs.py

```diff
@@ -48,53 +55,61 @@ def peek(self):
 
 class JobScheduler(threading.Thread):
-
     """Calls jobs assigned to it in steady intervals.
 
     JobScheduler is a thread that keeps track of Jobs and calls them every
-    X seconds, where X is a property of the Job. It maintains jobs in a
-    priority queue, where the next job to be called is always the first
-    item.
-    Thread safety is maintained with a mutex that is released during long
-    operations, so methods add_job and clear_jobs can be safely called from
-    the main thread.
+    X seconds, where X is a property of the Job.
 
-    """
-
-    min_reaction_time = 30.0  # seconds
-    """How often should scheduler checks for changes in the job list."""
+    Thread safety is ensured with an internal mutext.
 
+    It runs forever until the :attr:`stopping` event is set using the
+    :meth`stop` method.
+    """
     def __init__(self, bot):
         """Requires bot as argument for logging."""
         threading.Thread.__init__(self)
         self.bot = bot
-        self._jobs = PriorityQueue()
-        # While PriorityQueue it self is thread safe, this mutex is needed
-        # to stop old jobs being put into new queue after clearing the
-        # queue.
+        self.stopping = threading.Event()
+        self._jobs = []
         self._mutex = threading.Lock()
-        # self.cleared is used for more fine grained locking.
-        self._cleared = False
 
     def add_job(self, job):
         """Add a Job to the current job queue."""
-        self._jobs.put(job)
+        with self._mutex:
+            self._jobs.append(job)
 
     def clear_jobs(self):
         """Clear current Job queue and start fresh."""
-        if self._jobs.empty():
-            # Guards against getting stuck waiting for self._mutex when
-            # thread is waiting for self._jobs to not be empty.
-            return
         with self._mutex:
-            self._cleared = True
-            self._jobs = PriorityQueue()
+            self._jobs = []
+
+    def stop(self):
+        """Ask the job scheduler to stop.
+
+        The scheduler thread will stop its loop over jobs to process, but it
+        won't join the thread, or clear its queue - this has to be done
+        separatly by the calling thread.
+        """
+        self.stopping.set()
 
     def run(self):
-        """Run forever."""
-        while True:
+        """Run forever until :attr:`stopping` event is set."""
+        while not self.stopping.is_set():
             try:
-                self._do_next_job()
+                now = time.time()
+
+                # Collect ready jobs by now
+                for job in self._get_ready_jobs(now):
+                    self._run_job(job)
+
+                # Wait up to a second
+                time_spent = time.time() - now
+                wait_time = max(0, 1 - time_spent)
+                if wait_time:
+                    time.sleep(wait_time)
+            except KeyboardInterrupt:
+                # Do not lock on KeyboardInterrupt
```

Comment:
This at least helps with the previously-hidden ignored ^C I found when testing #1534, mentioned in https://github.com/sopel-irc/sopel/issues/1478#issuecomment-480043097, right? :+1:

(Note made here instead of next line so I could suggest a change to the comment, as usual…)
```suggestion
                # Do not block on KeyboardInterrupt
```

Reply:
It should help, but I noticed while testing that it doesn't completly remove the problem. I think there are other parts of the bot that don't like `^C`. I guess our multiple efforts will pay in the end.